### ISSUE TITLE
Npgsql throws 'A command is already in progress' in case of multiple async commands within one open connection.

### ISSUE THREAD

**Issue by melanore** [demetrixbio/FSharp.Data.Npgsql#32]
In certain cases, given open connection with a transaction, and multiple asynchronous dbcommands executed against it, - some command will undeterministically fail with 'A command is already in progress' error thrown by Npgsql library. After some investigation, it appeared, that this issue can occur in some cases, if DbReader is not being disposed. Below you can find answer from Npgsql library maintainer regarding similar use case:
https://stackoverflow.com/questions/36283841/c-sharp-winforms-npgsql-3-0-5-an-operation-already-in-progress-error-when-tryi
https://github.com/npgsql/npgsql/issues/1248
https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/284

How to reproduce
A bit tricky. Issue is not 100% reproduceable, I assume this has something to do with the case, when second dbCommand is executed on a different thread.

- Given asynchronous function
- OpenAsync dbConnection
- Call a function that will asynchronously execute query 1 against given connection
- Call a function that will asynchronously execute query 2 against given connection

Expected:
Both queries were successfully executed

Actual:
Second query will fail with 'A command is already in progress: Query 1'.

**Comment by melanore** [comment]
I've linked th PR which addresses the issue. Also, I've updated Npgsql library to 4.0.5 version, as they fixed another bug that was affecting batched inserts using Npgsql type providers. 
More details below:
https://github.com/npgsql/npgsql/issues/2329

**Comment by dmitry-a-morozov** [comment]
Great. I'll have a look on weekend.

**Comment by melanore** [comment]
Thanks!

**Comment by dmitry-a-morozov** [comment]
What problem do you try to solve?
My understanding that Npgsql/PostgreSQL as oppose to SQL Server conceptually doesn't support Multiple Active Result Sets a.k.a [MARS](https://docs.microsoft.com/en-us/sql/relational-databases/native-client/features/using-multiple-active-result-sets-mars?view=sql-server-2017).
So no matter what you do in type provider this feature is by design.

**Comment by melanore** [comment]
In this case I don't think there is a need in MARS.
The usecase is next:

`
let deleteStatement transaction = 
    async {
        use cmd = Db.CreateCommand<"DELETE sql">(transaction.Connection, transaction)
        return! cmd.AsyncExecute() |> Async.Ignore
    }

let insertStatement dtos transaction =
    async {
        let t = new Db.dbo.Tables.myTable()
        dtos
        |> List.iter (fun p ->
                let row = t.NewRow(x=p.x, y=p.y)
                t.Rows.Add row)
        let result = t.Update(transaction.Connection, transaction = transaction)
        return t.Rows |> Seq.map toDto |> List.ofSeq
    }

let serviceFunction dtos = 
    async {
        use! connection = asyncOpenConnection()
        use transaction = connection.BeginTransaction()
        do! deleteStatement transaction 
        let! result = insertStatement dtos 
        transaction.Commit()
        return result
    }
`
Basically, async functions, that execute multiple queries one by one with await against single open connection with transaction, sometimes throw 'operation is already in progress' exception.
There was one mentioning from Npgsql maintainer, that this can be caused by not disposed DbReaders.
https://stackoverflow.com/questions/36283841/c-sharp-winforms-npgsql-3-0-5-an-operation-already-in-progress-error-when-tryi
Upon checking type provider code, it appeared, that in some cases DbReaders were indeed no disposed, and, once fixed, - the bug stopped occurring within our app. Also, I've tested all other places for possible regression, but have noticed any (type provider tests also pass).

I hope this answers your question.

**Comment by dmitry-a-morozov** [comment]
Where is ```DbReader``` is coming from in the code above? There are no ```SELECT ...``` statements.

**Comment by dmitry-a-morozov** [comment]
Also I will follow up with you on PR later. I want to clearly understand the problem first.

**Comment by melanore** [comment]
In the first delete DbCommand, there is a 'RETURNING' statement, so that query in fact returns several columns as a result. Anyway, - you can find usage of DbReader in PR. I think, - disposing a reader is interpreted by Npgsql as a hint, that command is closed, according to Shay Rojansky.

Thanks for your time.

**Comment by dmitry-a-morozov** [comment]
So there is no concurrent execution in this case, right?
It's just a db resource (db reader) left not being cleaned-up.
So why not make sure it's disposed before proceeding forward?

Max, I don't mean to be annoying but if you want me to help I want to get to the bottom of it.
Feel free to merge PR anytime.
Here is some food for thoughts
https://github.com/fsprojects/FSharp.Data.SqlClient/issues/29

Also now output types are inaccurate. You expose computation like type ```seq<_>``` backed by already materialized data structure ```list<_>```. Users do not expect that.
Also would be nice to have unit tests. Because if you'll have correct unit tests you'll see that the PR doesn't fix the issue at all. It's only reduce probability of it.     
You can build repeatable tests using something like [WAITFOR](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/waitfor-transact-sql?view=sql-server-2017).

**Comment by daz10000** [comment]
Looking at the PR and the usecase,  I think yes, there is effectively concurrency due to the  reader remaining attached to the connection when the second instance uses it, and the PR does ensure the reader is closed by the time async completes execution.  The current master code would leave that read around for an indefinite preiod of time.  I made some comments too on the actual PR.  I do think the issue of the seq< _> semantics is a good one and not sure what a good solution is.  Ideally it implements IEnumerable rather than seq specifically

**Comment by dmitry-a-morozov** [comment]
I agree that eagerly evaluating underlying cursor reduces possibility of having more than one concurrent reader open. If you accept this change than output type should change too to accurately reflect reality. I mean instead of ```seq<_>```/```Async<seq<_>>``` methods ```Execute```/```AsyncExecute``` should return  ```list<_>```/```Async<list<_>>```.

**Comment by melanore** [comment]
Execute/AsyncExecute explicitly return ```list<_>```/```Async<list<_>>``` now. The change is backwards compatible as list also implements IEnumerable<_>, thus Seq.map etc is codebases that were making use of type provider won't break.

One thing to consider is use of FSharp.Control.AsyncSeq for async sequences. This will enable async open/close of reader to be part of sequence workflow, as well as async evaluation of results using ```DbDataReader.ReadAsync() |> Async.AwaitTask```, handling control over materialization to consumer of AsyncExecute. However, this is a breaking change and is tricky to implement, so after a day of playing around with this idea I couldn't make it work :(
